brainiak/funcalign/fastsrm.py

```diff
@@ -0,0 +1,1014 @@
+        self.n_components = n_components
+        self.n_iter = n_iter
+        self.atlas = atlas
+
+        self.basis_list = None
+
+        if temp_dir is None:
+            if self.verbose == "warn" or self.verbose is True:
+                logger.warning("temp_dir has value None. "
+                               "All basis (spatial maps) and reconstructed "
+                               "data will therefore be kept in memory."
+                               "This can lead to memory errors when the "
+                               "number of subjects "
+                               "and/or sessions is large.")
+            self.temp_dir = None
+            self.low_ram = False
+
+        if temp_dir is not None:
+            if not os.path.exists(os.path.join(temp_dir, "fastsrm")):
+                os.mkdir(os.path.join(temp_dir, "fastsrm"))
+            self.temp_dir = os.path.join(temp_dir, "fastsrm")
+
+            # Remove files in temp folder
+            paths = glob.glob(os.path.join(self.temp_dir, "*.npy"))
+            for path in paths:
+                os.remove(path)
+
+            self.low_ram = low_ram
+
+    def fit(self, imgs):
+        """Computes basis across subjects from input imgs
+
+        Parameters
+        ----------
+
+        imgs : array of str, shape=[n_subjects, n_sessions]
+            Element i, j of the array is a path to the data of subject i
+            collected during session j.
+            Data are loaded with numpy.load and expected
+            shape is [n_timeframes, n_voxels]
+            n_timeframes and n_voxels are assumed to be the same across
+            subjects
+            n_timeframes can vary across sessions
+            Each voxel's timecourse is assumed to have mean 0 and variance 1
+
+        Returns
+        -------
+        self : object
+           Returns the instance itself. Contains attributes listed
+           at the object level.
+        """
+
+        if self.temp_dir is not None:
+            # Remove former basis in temp folder
+            paths = glob.glob(os.path.join(self.temp_dir, "*.npy"))
+            for path in paths:
+                os.remove(path)
+
+        check_imgs(imgs)
+
+        if self.verbose is True:
+            n_subjects, n_sessions = imgs.shape
+            logger.info("Fitting using %i subjects and %i sessions per subject"
+                        % (n_subjects, n_sessions))
+            logger.info("[FastSRM.fit] Reducing data")
+
+        reduced_data = reduce_data(
+            imgs,
+            atlas=self.atlas,
+            n_jobs=self.n_jobs,
+            low_ram=self.low_ram,
+            temp_dir=self.temp_dir
+        )
+
+        if self.verbose is True:
+            logger.info("[FastSRM.fit] Finds shared "
+                        "response using reduced data")
+
+        shared_response_list = fast_srm(
+            reduced_data,
+            n_iter=self.n_iter,
+            n_components=self.n_components,
+        )
+
```

Comment:
This fast_srm function is basically DetSRM. Since this part does not use parallelization anyway, could you re-use the functions (DetSRM and SRM) in srm.py to do this fitting?

Reply:
fast_srm does have one thing that DetSRM does not have which is the "low_ram" mode. In some cases even the reduced dataset might not hold in memory (when the number of sessions is very big and the number of regions in the atlas are big) in which case it can be useful to write reduced data on disk. 
Typically this "low_ram" mode is useful when the number of voxels is a lot less than the number of sessions x number of subjects x number of reduced components.
This functionality is not mentionned in the paper and I did not use it in the experiments. I could therefore get read of it but as datasets will grow bigger it might be useful at some point.
Another option would be to call DetSRM if low_ram mode is not used and to use the current implementation if low_ram mode is used.

Reply:
I think it is better to call DetSRM/SRM if low_ram is not used and to use the current implementation if low_ram mode is used. An advantage of this is that when low_ram is not used, you can also apply ProbSRM (or any other algorithms similar to SRM).